pkg/pillar/types/global.go

```diff
@@ -103,6 +110,237 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+		TriStateDefault: defaultTriState,
+	}
+	if agentSetting {
+		configMap.AgentSettings[key][agentSettingKey] = configItem
+	}
+	if globalSetting {
+		configMap.GlobalSettings[key] = configItem
+	}
+	log.Debugf("Added tristate item %s", key)
+}
+
+type ConfigItemValue struct {
+	Key      string
+	ItemType ConfigItemType
+
+	IntValue      uint32
+	StringValue   string
+	BoolValue     bool
+	TriStateValue TriState
+}
+
+type ConfigItemValueMap struct {
+	GlobalSettings map[string]ConfigItemValue
+	AgentSettings  map[string]map[string]ConfigItemValue
+}
+
+func GetDefaultValueMap(configMap *ConfigItemSpecMap) ConfigItemValueMap {
+	var valueMap ConfigItemValueMap
+	var item ConfigItemValue
+	for key, value := range configMap.GlobalSettings {
+		if value.ItemType == CONFIGITEM_TYPE_BOOL {
+			item.BoolValue = value.BoolDefault
+			valueMap.GlobalSettings[key] = item
+		} else if value.ItemType == CONFIGITEM_TYPE_INT {
+			item.IntValue = value.IntDefault
+			if item.IntValue > value.IntMax {
+				item.IntValue = value.IntMax
+			} else if item.IntValue < value.IntMin {
+				item.IntValue = value.IntMin
+			}
+			valueMap.GlobalSettings[key] = item
+		} else if value.ItemType == CONFIGITEM_TYPE_STRING {
+			item.StringValue = value.StringDefault
+			valueMap.GlobalSettings[key] = item
+		} else if value.ItemType == CONFIGITEM_TYPE_TRISTATE {
+			item.TriStateValue = value.TriStateDefault
+			valueMap.GlobalSettings[key] = item
+		}
+	}
+	for agentName, perAgentMap := range configMap.AgentSettings {
+		for agentSettingName, value := range perAgentMap {
+			if value.ItemType == CONFIGITEM_TYPE_BOOL {
+				item.BoolValue = value.BoolDefault
+				valueMap.AgentSettings[agentName][agentSettingName] = item
+			} else if value.ItemType == CONFIGITEM_TYPE_INT {
+				item.IntValue = value.IntDefault
+				if item.IntValue > value.IntMax {
+					item.IntValue = value.IntMax
+				} else if item.IntValue < value.IntMin {
+					item.IntValue = value.IntMin
+				}
+				valueMap.AgentSettings[agentName][agentSettingName] = item
+			} else if value.ItemType == CONFIGITEM_TYPE_STRING {
+				item.StringValue = value.StringDefault
+				valueMap.AgentSettings[agentName][agentSettingName] = item
+			} else if value.ItemType == CONFIGITEM_TYPE_TRISTATE {
+				item.TriStateValue = value.TriStateDefault
+				valueMap.AgentSettings[agentName][agentSettingName] = item
+			}
+		}
+	}
+	return valueMap
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, item config.ConfigItem) error {
+	key := item.Key
+	configItem := ConfigItemValue{
+		Key:	key,
+	}
+	agentSetting := false
+	globalSetting := false
+	configSpecMap := specMap.GlobalSettings
+	if _, ok := specMap.GlobalSettings[key]; ok {
+		configSpecMap = specMap.GlobalSettings
+		globalSetting = true
+	} else {
+		components := strings.Split(key, ".")
+		if _, ok := specMap.AgentSettings[components[1]]; ok && (components[0] == "debug" || components[0] == "agent") {
```

Comment:
1. How do you know that the split results in at least two components?
2. This looks like it can allow setting any variable per agent, which we don't want. Just the two loglevels should be per agent. So defining an array of allowed per-agent names makes sense.

Reply:
One thing we discussed is - Going forward, for future agent settings, we move to the format:
   agent.<agentname>.<key string>
     - This would be the key for all agent settings..

We will handle the current "debug.<agentname>" as exceptions during transition.. 
But this code looks fishy neverthless..

Reply:
"So defining an array of allowed per-agent names makes sense." --> In one of the above comment, proposing to define the allowed keys as its own type and have Constant values for valid ones.

For your comment - SpecMap already has a map of all allowed keys - both Global keys as well as allowed per-agent keys.. So that should protect us from accepting invalid keys.

Reply:
Resolved 1